utils/s2n_array.c

```diff
@@ -51,28 +51,48 @@ static S2N_RESULT s2n_array_enlarge(struct s2n_array *array, uint32_t capacity)
 
 struct s2n_array *s2n_array_new(uint32_t element_size)
 {
-    struct s2n_blob mem = {0};
+    struct s2n_array *array = s2n_array_new_with_capacity(element_size, S2N_INITIAL_ARRAY_SIZE);
+    PTR_ENSURE_REF(array);
+
+    return array;
+}
+
+struct s2n_array *s2n_array_new_with_capacity(uint32_t element_size, uint32_t capacity)
+{
+    struct s2n_blob mem = { 0 };
     PTR_GUARD_POSIX(s2n_alloc(&mem, sizeof(struct s2n_array)));
 
-    struct s2n_array *array = (void *) mem.data;
+    DEFER_CLEANUP(struct s2n_array *array = (void *) mem.data, s2n_array_free_p);
 
-    *array = (struct s2n_array) {.mem = {0}, .len = 0, .element_size = element_size};
+    PTR_GUARD_RESULT(s2n_array_init_with_capacity(array, element_size, capacity));
```

Comment:
I know this isn't a big concern, but I'm still worried that it's theoretically possible to leak here. What if alloc fails, but after allocating memory?

What happens if you DEFER_CLEANUP both mem and array? Does that cause an error? Or do we need to ZERO_TO_DISABLE_DEFER_CLEANUP(mem) after array is declared?

None of these are likely / testable failure cases with the current code though. Like, s2n_alloc and s2n_array_init_with_capacity really shouldn't fail.

Reply:
I added the ZERO_TO_DISABLE_DEFER_CLEANUP(mem) after array is declared.